internal/preflight/preflight.go

```diff
@@ -0,0 +1,99 @@
+package preflight
+
+import (
+	"context"
+	"fmt"
+
+	kmmv1beta1 "github.com/qbarrand/oot-operator/api/v1beta1"
+	"github.com/qbarrand/oot-operator/internal/auth"
+	"github.com/qbarrand/oot-operator/internal/module"
+	"github.com/qbarrand/oot-operator/internal/registry"
+
+	"k8s.io/apimachinery/pkg/types"
+	ctrlruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	VerificationStatusReasonBuildConfigPresent = "Verification successful, all driver-containers have paired BuildConfigs in the recipe"
+	VerificationStatusReasonNoDaemonSet        = "Verification successful, no driver-container present in the recipe"
+	VerificationStatusReasonUnknown            = "Verification has not started yet"
+	VerificationStatusReasonVerified           = "Verification successful, this Module would be verified again in this Preflight CR"
+)
+
+//go:generate mockgen -source=preflight.go -package=preflight -destination=mock_preflight_api.go
+
+type PreflightAPI interface {
+	PreflightUpgradeCheck(ctx context.Context, mod *kmmv1beta1.Module, kernelVersion string) (bool, string)
+}
+
+func NewPreflightAPI(
+	client client.Client,
+	registryAPI registry.Registry,
+	kernelAPI module.KernelMapper) PreflightAPI {
+	return &preflight{
+		registryAPI: registryAPI,
+		kernelAPI:   kernelAPI,
+		client:      client,
+	}
+}
+
+type preflight struct {
+	client      client.Client
+	registryAPI registry.Registry
+	kernelAPI   module.KernelMapper
+}
+
+func (p *preflight) PreflightUpgradeCheck(ctx context.Context, mod *kmmv1beta1.Module, kernelVersion string) (bool, string) {
+	mapping, err := p.kernelAPI.FindMappingForKernel(mod.Spec.ModuleLoader.Container.KernelMappings, kernelVersion)
+	if err != nil {
+		return false, fmt.Sprintf("Failed to find kernel mapping in the module %s for kernel version %s", mod.Name, kernelVersion)
+	}
+
+	osConfig := module.NodeOSConfig{KernelFullVersion: kernelVersion}
+	mapping, err = p.kernelAPI.PrepareKernelMapping(mapping, &osConfig)
+	if err != nil {
+		return false, fmt.Sprintf("Failed to substitute template in kernel mapping in the module %s for kernel version %s", mod.Name, kernelVersion)
+	}
+
+	return p.verifyImage(ctx, mapping, mod, kernelVersion)
+}
+
+func (p *preflight) verifyImage(ctx context.Context, mapping *kmmv1beta1.KernelMapping, mod *kmmv1beta1.Module, kernelVersion string) (bool, string) {
```

Comment:
Could this just return an error?

Reply:
error won't help us. This should return whether the image is verified, and also the message that describes the reason for verification or not verification

Reply:
We could do:
- image verified: depends on whether `err` is nil or not
- reason: the error string
WDYT?

Reply:
I would prefer not to: error means that something went wrong with the flow, while in preflight case there is nothing wrong with the flow, the function  just returns the status of verification.